File: data/derived_cot/rq2_eval/translation/r1/_converted_code/py_to_cpp/Warehouse.cpp
#include <unordered_map>
#include <string>
#include <variant>

struct Product {
    std::string name;
    int quantity;
};

struct Order {
    int product_id;
    int quantity;
    std::string status;
};

class Warehouse {
private:
    std::unordered_map<int, Product> inventory;
    std::unordered_map<int, Order> orders;

public:
    void add_product(int product_id, const std::string& name, int quantity) {
        auto it = inventory.find(product_id);
        if (it == inventory.end()) {
            inventory[product_id] = Product{name, quantity};
        } else {
            it->second.quantity += quantity;
        }
    }

    void update_product_quantity(int product_id, int quantity) {
        auto it = inventory.find(product_id);
        if (it != inventory.end()) {
            it->second.quantity += quantity;
        }
    }

    int get_product_quantity(int product_id) {
        auto it = inventory.find(product_id);
        if (it == inventory.end()) {
            return -1;
        }
        return it->second.quantity;
    }

    bool create_order(int order_id, int product_id, int quantity) {
        auto it = inventory.find(product_id);
        int available = 0;
        if (it != inventory.end()) {
            available = it->second.quantity;
        }
        if (available >= quantity) {
            if (it != inventory.end()) {
                it->second.quantity -= quantity;
            }
            orders[order_id] = Order{product_id, quantity, "Shipped"};
            return true;
        }
        return false;
    }

    bool change_order_status(int order_id, const std::string& status) {
        auto it = orders.find(order_id);
        if (it == orders.end()) {
            return false;
        }
        it->second.status = status;
        return true;
    }

    std::variant<std::string, bool> track_order(int order_id) {
        auto it = orders.find(order_id);
        if (it == orders.end()) {
            return false;
        }
        return it->second.status;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/r1/cpp_to_py/code_output/Warehouse.py
class Warehouse:
    def __init__(self):
        self._inventory = {}   # int -> dict[str, str]  keys: "name", "quantity"
        self._orders = {}      # int -> dict[str, str]  keys: "product_id", "quantity", "status"

    def add_product(self, product_id: int, name: str, quantity: int) -> None:
        if product_id not in self._inventory:
            self._inventory[product_id] = {
                "name": name,
                "quantity": str(quantity)
            }
        else:
            product = self._inventory[product_id]
            current_qty = int(product["quantity"])
            product["quantity"] = str(current_qty + quantity)

    def update_product_quantity(self, product_id: int, quantity: int) -> None:
        if product_id in self._inventory:
            product = self._inventory[product_id]
            current_qty = int(product["quantity"])
            product["quantity"] = str(current_qty + quantity)

    def get_product_quantity(self, product_id: int) -> int:
        if product_id in self._inventory:
            return int(self._inventory[product_id]["quantity"])
        return 0  # matches C++ returning false (implicit int 0)

    def create_order(self, order_id: int, product_id: int, quantity: int) -> bool:
        available = self.get_product_quantity(product_id)
        if available >= quantity:
            self.update_product_quantity(product_id, -quantity)
            self._orders[order_id] = {
                "product_id": str(product_id),
                "quantity": str(quantity),
                "status": "Shipped"
            }
            return True
        return False

    def change_order_status(self, order_id: int, status: str) -> bool:
        if order_id in self._orders:
            self._orders[order_id]["status"] = status
            return True
        return False

    def track_order(self, order_id: int) -> str:
        if order_id in self._orders:
            return self._orders[order_id]["status"]
        return ""

    # Note: In C++ these return const references, so the caller cannot modify.
    # Here we return the actual dicts for simplicity (common Python practice).
    def orders(self):
        return self._orders

    def inventory(self):
        return self._inventory